src/node/blockstorage.cpp

```diff
@@ -397,6 +404,16 @@ bool BlockManager::IsBlockPruned(const CBlockIndex* pblockindex)
     return (m_have_pruned && !(pblockindex->nStatus & BLOCK_HAVE_DATA) && pblockindex->nTx > 0);
 }
 
+const CBlockIndex* GetFirstStoredBlock(const CBlockIndex* start_block) {
+    AssertLockHeld(::cs_main);
+    assert(start_block);
+    const CBlockIndex* last_block = start_block;
+    while (last_block->pprev && (last_block->pprev->nStatus & BLOCK_HAVE_DATA)) {
+        last_block = last_block->pprev;
+    }
+    return last_block;
```

Comment:
Would be nice if there was a way to annotate the returned value is never nullptr

Reply:
I guess this can be achieved by deleting the move/copy constructor of `CBlockIndex` and then returning a reference instead of pointer.

Reply:
> I guess this can be achieved by deleting the move/copy constructor of `CBlockIndex` and then returning a reference instead of pointer.

interesting